### PR TITLE
Tree and graphs Questions

### PR DESCRIPTION
**Pull request by anilhariparamhans**

### CODE REVIEW

**Comment by anilhariparamhans** [comment]
@dimplenaresh1  I added the code for CTCI 8.2 Robot in Grid
Please review this pull request.

File: CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		coutNumWayshV1(R, C, r + 1, c);
+
+	if (c < C)
+		coutNumWayshV1(R, C, r, c + 1);
+}
+
+_API size_t coutNumWaysV1(int r, int c)
+{
+	coutNumWayshV1(r, c, 0, 0);
+	return ans1;
```
Comment (dimplenaresh1) [inline CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp:49]: Question is: Design an algorithm to find a path for the robot from the top left to
the bottom right.
So we have to return the path, not number of possible ways.

File: CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
+#include "robotheader.h"
+#include <vector>
+
+size_t ans1 = 0;
+void coutNumWayshV1(const int R, const int C, int r, int c)
+{
```
Comment (dimplenaresh1) [inline CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp:24]: Where have you taken care of invalid cells where robot can not step?

File: CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	return w;
+}
+
+_API size_t coutNumWaysV2(int r, int c)
+{
+	std::vector<std::vector<int>> mp(r, std::vector<int>(c, 0));
+
+	coutNumWayshV2(mp, r-1, c-1);
+
+	return mp[r - 1][c - 1];
```
Comment (dimplenaresh1) [inline CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp:96]: we have to return the path from top left to bottom right.

File: CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp (hunk 1 above)
Comment (anilhariparamhans) [reply CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp:49]: Yes, correct. I added the function to get path.

File: CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp (hunk 2 above)
Comment (anilhariparamhans) [reply CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp:24]: I forgot to add the logic for invalid cells.

I added this logic.

Please review.

File: CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp (hunk 3 above)
Comment (anilhariparamhans) [reply CrackingCodingInterview/vAnil/08 Recursion and Dynamic Programming/02 A_RobotInGrid/robot.cpp:96]: Yes. I added the function to get path. 

Please review.

**Comment by anilhariparamhans** [comment]
@dimplenaresh1 I incorporated all the review comments.
Please review the code changes.

**Review by dimplenaresh1 — APPROVED** [review]

**Pull request merged**